syntax = "proto3";
import "runtime.proto";


// Runs alongside each trainer to perform gating function averaging every now and then. Read more: client/averaging.py
service DecentralizedAveraging {
  rpc rpc_group_allreduce(MessageToLeader) returns (stream MessageFromLeader);  // assemble a group and run all-reduce
  rpc rpc_aggregate_part(AveragingData) returns (AveragingData);  // send my local shard => get aggregated shard
}

message MessageToLeader {
  bytes scheme_hash = 1;        // I would like to average this kind of data (used for sanity check)
  string leader_endpoint = 3;  // This is how a (potential) follower perceives the leader (used for sanity check)
  float expiration = 4;         // Follower would like to **begin** all_reduce by this point in time
}

enum MessageCode {
  // response to join request
  ACCEPTED = 0;              // "I accept you in my group / as a freeloader."
  NOT_A_LEADER = 1;          // "I am not a group a leader. Go ask my leader instead."
  ALREADY_RUNNING = 2;       // "My group has already began merging. Here's the group leader."
  NOT_LOOKING_FOR_GROUP = 3; // "I'm not available at the moment. Please, get lost."
  BAD_EXPIRATION_TIME = 4;   // "I will not accept you. I cannot guarantee that we begin before you expire."
  BAD_SCHEME_HASH = 5;       // "I will not accept you. I am not averaging the samy type of tensors as you."

  // message from leader when finding peers
  GROUP_HEARTBEAT = 6;       // "We're still in gathering. Please, stay with us."
  BEGIN_ALLREDUCE = 7;       // "We can begin allreduce now. These are your peers."
  GROUP_DISMISSED = 8;       // "The group is closed. Go find another group."

  // when averaging
  ALLREDUCE_FINISHED = 9;   // "We finished allreduce successfully, you can use the results."
  ALLREDUCE_FAILED = 10;     // "Something went wrong during allreduce. Your results may be wrong."
}


message MessageFromLeader {
  MessageCode code = 1;
  bytes group_id = 2;        // a unique identifier of this group, only valid until allreduce is finished/failed
  string suggested_leader = 3;  // if peer is already in a group, it'll provide us with an endpoint of its leader
  string follower_endpoint = 4;  // This is how a group leader sees its follower (used for sanity check)
  repeated string group_endpoints = 5;  // a sequence of peers, each responsible for one shard during averaging
}

message AveragingData {
  bytes group_id = 1;        // a unique group identifier, same as in MessageFromLeader
  bytes scheme_hash = 2;     // a hash of your key, tensor shape, and part index (used for sanity check)
  Tensor tensor = 3;         // either peer's local tensor part (rpc input) or group average of this part (rpc output)
  string error = 4;          // in case of protocol violation, this will be the error message
}
